Moved atomic decrement outside the lock

rkothiya opened the pull request:
Fixes: #1429

Change-Id: I37144eba3e25d9262bb2f4aace5695f6e4285063
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

Commit:
Moved atomic decrement outside the lock

Fixes: #1429

Change-Id: I37144eba3e25d9262bb2f4aace5695f6e4285063
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

rkothiya commented on xlators/performance/write-behind/src/write-behind.c:
```diff
@@ -3191,15 +3191,13 @@ struct volume_options options[] = {
-        .tags = {"write-behind"},
-        .description = "Enable/disable write-behind pass-through"
-    },
+    {.key = {"pass-through"},
```
These are clang changes I did not do this formatting.

mohit84 commented on xlators/performance/write-behind/src/write-behind.c:
```diff
@@ -2512,12 +2512,12 @@ wb_mark_readdirp_end(xlator_t *this, inode_t *directory)
     if (!wb_directory_inode)
         return;
 
+    readdirps = GF_ATOMIC_DEC(wb_directory_inode->readdirps);
```
I think we need to change in wb_mark_readdirp_start also, readdirps is atomic variable we don;t need to access in LOCK/UNLOCK

Commit:
Moved atomic decrement outside the lock

Implemented the review comment to move the atomic increment out
of the lock and unlock in the wb_mark_readdirp_start function.

Fixes: #1429

Change-Id: I8a2cffe39109efa06e9a2b98026f9dbe78f179d1
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

rkothiya commented:
/recheck smoke

rkothiya commented:
/run regression

xhernandez reviewed (commented):
Have you fully traced all the implications of updating the atomic variable outside the lock ?

At first sight this atomic seems related to the `invalidate_list` list. Are you sere there's no dependency that can cause them to be out of sync if not updated *both* atomically (using the mutex) ?

Commit:
Moved atomic decrement outside the lock

Moved atomic functions out of the mutex in wb_set_invalidate()
to keep the related code consistent with other changes.

Fixes: #1429

Change-Id: Ic60b3be3b14961fb371aa0df11af0e2ed47b9f5d
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

rkothiya commented:
IMHO it should not cause any problem. I have moved atomic functions out of the mutex in wb_set_invalidate() also to keep the related code consistent with other changes.

rkothiya commented:
/run regression

xhernandez commented:
Given the surprises we have got with #1380, which seemed to me much more safe than this one, I think we shouldn't merge this patch unless we see a clear performance improvement that justifies the risks and effort to review and fix any potential future issue. Otherwise I would prefer to leave the code as it is now.

mykaul commented:
> Given the surprises we have got with #1380, which seemed to me much more safe than this one, I think we shouldn't merge this patch unless we see a clear performance improvement that justifies the risks and effort to review and fix any potential future issue. Otherwise I would prefer to leave the code as it is now.

I think the lesson here is different - CI was failing, on this patch, several times. While issues were opened (https://github.com/gluster/glusterfs/issues/1975 and https://github.com/gluster/glusterfs/issues/1982 ) we still merged this change. We probably should not have.

When I originally developed a similar patch, I saw a reduction in the number of calls to lock. This isn't a direction indication of a performance improvement of course. I also saw less calls going into that part of the code when looking at it with 'perf'.

xhernandez commented:
> I think the lesson here is different - CI was failing, on this patch, several times. While issues were opened (#1975 and #1982 ) we still merged this change. We probably should not have.

I agree. I overlooked it. I'm sorry for that...

However it doesn't change the fact that I see more risk in this change than the other one. Something doesn't "feel" right to me here, but I haven't detected any issue yet. I would have to spend more time reviewing this patch to be confident enough with it or to detect some race. I'm reluctant to spend so much time on this if the change doesn't really provide any benefit.

mykaul commented:
@xhernandez  - Isn't the original code also prone to a race? We fetch refcount within the lock, but then, based on that value (let's assume it was 0), we act - outside the lock - and potentially destroy that fd - what if someone else incremented it in between?
https://github.com/gluster/glusterfs/blob/22910dcc1046e3d7eeaba89b05093a6fcc37e172/libglusterfs/src/fd.c#L559

xhernandez commented:
> @xhernandez - Isn't the original code also prone to a race? We fetch refcount within the lock, but then, based on that value (let's assume it was 0), we act - outside the lock - and potentially destroy that fd - what if someone else incremented it in between?
> https://github.com/gluster/glusterfs/blob/22910dcc1046e3d7eeaba89b05093a6fcc37e172/libglusterfs/src/fd.c#L559

fd's can't be used without a reference (otherwise it's a bug an will probably lead to a crash sooner or later). So if there's a thread doing an unref and another thread doing a ref, both threads must own a reference, so refcount before calling unref must be, at least, 2, which means that the unref will never have refcount = 0 and the fd won't be destroyed.

Atomic ref counting is a very well known pattern and, if used correctly, it's safe. In this patch the atomic variable doesn't seem a pure ref counter. It also seems highly coupled to a list that is checked outside a lock, so if the valid configurations of both variables need to follow some restrictions, not updating/checking them inside a lock to ensure atomicity may lead to unexpected races.
